pilot/pkg/serviceregistry/aggregate/controller.go

```diff
@@ -228,11 +228,25 @@ func (c *Controller) GetProxyServiceInstances(node *model.Proxy) ([]*model.Servi
 	// It doesn't make sense for a single proxy to be found in more than one registry.
 	// TODO: if otherwise, warning or else what to do about it.
 	for _, r := range c.GetRegistries() {
+		// Skip the service registry when we know there won't be a match
+		// because the proxy is in a different cluster. We can't trust
+		// the default service registry because its always named `Kubernetes`
+		// and won't match the CLUSTER_ID label from the proxies.
+		if r.Cluster() != "" &&
+			r.Cluster() != string(serviceregistry.Kubernetes) &&
+			node.Metadata != nil &&
+			node.Metadata.ClusterID != "" &&
+			node.Metadata.ClusterID != r.Cluster() {
+			log.Debugf("GetProxyServiceInstances(): not checking cluster %v: proxy %v is associated with cluster %v",
+				r.Cluster(), node.ID, node.Metadata.ClusterID)
+			continue
```

Comment:
So if `r.Cluster == Kubernetes` and `node.Cluster == MyOtherCluster`, this will NOT be skipped? Isn't that what we were trying to avoid?

Reply:
Yes, that is one of the cases we were trying to avoid. ClusterID is overloaded to mean both the registry type (k8s, consul) and the name (which cluster). I didn't want to fix that issue holistically in this PR (see / https://github.com/istio/istio/issues/22093). On further thought though we could filter 100% of the cases if we had some special handling above that is less intrusive (see skipSearchingRegistryForProxy above).